js/src/shared/redux/providers/walletActions.js

```diff
@@ -0,0 +1,520 @@
+          confirmedBy: []
+        }));
+      })
+      .then((logs) => {
+        return logs.sort((logA, logB) => {
+          const comp = logA.blockNumber.comparedTo(logB.blockNumber);
+
+          if (comp !== 0) {
+            return comp;
+          }
+
+          return logA.transactionIndex.comparedTo(logB.transactionIndex);
+        });
+      })
+      .then((confirmations) => {
+        if (confirmations.length === 0) {
+          return confirmations;
+        }
+
+        // Only fetch confirmations for operations not
+        // yet confirmed (ie. not yet a transaction)
+        if (transactions) {
+          const operations = transactions
+            .filter((t) => t.operation)
+            .map((t) => t.operation);
+
+          return confirmations.filter((confirmation) => {
+            return !operations.includes(confirmation.operation);
+          });
+        }
+
+        return confirmations;
+      });
+  } else {
+    const { confirmations } = wallet;
+    const nextConfirmations = confirmations
+      .filter((conf) => _operations.includes(conf.operation));
+
+    promise = Promise.resolve(nextConfirmations);
+  }
+
+  return promise
+    .then((confirmations) => {
+      if (confirmations.length === 0) {
+        return confirmations;
+      }
+
+      const uniqConfirmations = Object.values(
+        confirmations.reduce((confirmations, confirmation) => {
+          confirmations[confirmation.operation] = confirmation;
+          return confirmations;
+        }, {})
+      );
+
+      const operations = uniqConfirmations.map((conf) => conf.operation);
+
+      return Promise
+        .all(operations.map((op) => fetchOperationConfirmations(contract, op, owners)))
+        .then((confirmedBys) => {
+          uniqConfirmations.forEach((_, index) => {
+            uniqConfirmations[index].confirmedBy = confirmedBys[index];
+          });
+
+          return uniqConfirmations;
+        });
+    })
+    .then((confirmations) => {
+      const prevConfirmations = wallet.confirmations || [];
+      const nextConfirmations = prevConfirmations
+        .filter((conA) => !confirmations.find((conB) => conB.operation === conA.operation))
+        .concat(confirmations)
+        .map((conf) => ({
+          ...conf,
+          pending: false
+        }));
+
+      return {
+        key: UPDATE_CONFIRMATIONS,
+        value: nextConfirmations
+      };
+    });
+}
+
+function fetchOperationConfirmations (contract, operation, owners = null) {
+  if (!owners) {
+    console.warn('[fetchOperationConfirmations] try to provide the owners for the Wallet', contract.address);
+  }
+
+  const walletInstance = contract.instance;
+
+  const promise = owners
+    ? Promise.resolve({ value: owners })
+    : fetchWalletOwners(contract);
+
+  return promise
+    .then((result) => {
+      const owners = result.value;
+
+      return Promise
+        .all(owners.map((owner) => walletInstance.hasConfirmed.call({}, [ operation, owner ])))
```

Comment:
Why not `walletInstance.hasConfirmed([operation, owner])`? Is changing the `this` context to empty object a quirk?